### DISCUSSION TITLE
Multiplicative order algorithm

### DISCUSSION THREAD

**Discussion by jxu** [sympy/sympy#24422] [Q&A] [answered]
What algorithm is used for determining multiplicative order? https://github.com/sympy/sympy/blob/26f7bdbe3f860e7b4492e102edec2d6b429b5aaf/sympy/ntheory/residue_ntheory.py#L18-L58

I guess it is the LCM of the orders of the prime powers that n factors as. Also is the fact the order must divide totient(n) used?

**Comment by oscarbenjamin** [comment]
I don't know the answer to your question but it doesn't really sound like an issue to be fixed so I'm going to turn this issue into a discussion.

**Comment by asmeurer** [comment]
Is your point that we should be using algorithm that doesn't use prime factorization? Is there such an algorithm?

**Reply by jxu** [reply]
My question is if the algorithm uses the fact that order must be a divisor of phi(n) (Lagrange's theorem)

**Comment by jksuom** [comment, marked as answer]
> if the algorithm uses the fact that order must be a divisor of phi(n) 

Yes, it does; phi(n) is called `group_order`. It is computed on the lines 37 - 47 though probably it would be possible to use `totient(n)`. The algorithm works by splitting `group_order` into a product of prime power factors $p_i^{e_i}$. Then the order will be a product of powers $p_i^{d_i}$ where $d_i \le e_i$ for each i. The exponents $d = d_i$ are computed in the loop 51 - 57 (where $d_i$ is denoted `e - f + 1`) by starting from $d = e_i$ and decrementing $d$ by 1 as long as $\mathrm{pow}(a, \varphi(n)p_i^{d-e_i}, n) == 1$.
